### DISCUSSION TITLE
How to make the ghpages bot wait instead of cancelling?

### DISCUSSION THREAD

**Discussion by cadop** [community/community#39196] [Actions] [answered]
### Select Topic Area

Question

### Body

I have two jobs that run in an action.  One of them builds C++ documentation, the other builds python documentation.  At the end of each job, there is a deploy action that pushes to a ghpages. 

This seems to generate two different `pages build and deployment` actions.  However, it randomly chooses to either run both, or sometimes cancel one of them:
![image](https://user-images.githubusercontent.com/11399119/202036704-dbbb2b4a-188b-4d66-9ab7-dc474c285e0f.png)

The reason states a higher priority:
![image](https://user-images.githubusercontent.com/11399119/202036754-4cacabb8-f073-486d-bebb-04ed3b853df3.png)

Is there a way to make one of the jobs wait until the doc deployment finishes?

**Comment by airtower-luna** [comment, marked as answer]
It wouldn't make sense to make the second job wait: The second deployment will overwrite the first one anyway, so there's no point in finishing the first build. Okay, unless the second one will fail, I suppose. :sweat_smile: 

However you might want to consider using a [custom Pages workflow](https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site#publishing-with-a-custom-github-actions-workflow) for the deployment. That was introduced fairly recently, it allows you to deploy the site directly from the workflow that builds your documentation, without committing to a `ghpages` branch. For example you could create a job that `needs` the jobs that build the two documentation packages, retrieves them as artifacts, and deploys them together. :slightly_smiling_face:

**Reply by cadop** [reply]
I guess if I can't find a way to do this i'll learn the custom Pages, but was hoping to just modify what I have.  

Is it not possible or you are just saying you don't think it will work?  From the git side of it I don't think they should overwrite eachother since they are not touching the same folders. The docs are for two different folders, so one builds c++ docs, the other python docs.  Are you saying the automatic deploy action that is happening copies the entire branch that was pushed instead of doing the git push that was requested?

**Reply by airtower-luna** [reply]
> Are you saying the automatic deploy action that is happening copies the entire branch that was pushed instead of doing the git push that was requested?

Pretty much, yes. You do the push, and then Pages deploys *everything* from the latest commit on that branch. That's why I wrote it doesn't make sense to wait for the first build to finish, when the second one is going to deploy either the same or newer files.

**Reply by cadop** [reply]
If I push to the branch, and if it was just using git, I would end up with all my files on the branch correctly, then I would want it to build the pages.   So can I make the page deploys wait (and then only the second one run, after both pushes have been made to the branch)?

I'm not trying to argue/belabor the point, just trying to understand how pages/actions works (i've just been able to get some things to work, and lately this seems to have stopped working).

**Reply by cadop** [reply]
sorry to bother you. this is the part I want to make wait: 

> You do the push, and then Pages deploys everything from the latest commit on that branch.  

I have two jobs, I want them to be sequential:

 `Pull docs branch` -> `build C++ pages` -> `deploy c++ pages` -> `Pull docs branch` ->`build python pages` -> `deploy python pages`.  This is what I mean by making the job wait for pages to finish deploying. 

I can only get it to wait for the previous job to finish but is there a way to wait until that jobs' page deploy also finishes so I can start again (pulling the latest commit, including the just pushed c++ docs).

**Reply by airtower-luna** [reply]
If there's a push to the designated branch, Pages makes a deployment from the just-pushed state. By "state" I mean the *whole* checked-out branch, it doesn't care which commit changed what file. That's all.

With your docs builds pushing separately in quick succession, the first deployment gets cancelled because it would be a waste of time and processing power to finish it if there's going to be another deployment immediately after. Instead Pages just cancels the older deployment and deploys the latest version immediately.

If you want to get rid of the canceled job (good!), push both sets of documentation together. You could store the build results in artifacts, and have another job retrieve both artifacts and push (or use the custom deploy method right there, so you don't even need to bother pushing anything). Trying to make the doc build jobs wait would be very fiddly (API calls to check the deployment status for the pushed commit, sleep if it's not done yet, repeat, etc.) and the _only_ thing it'd achieve is slowing down your deployment (and wasting processing power), while the end result stays the same.

Or is your problem that one of the doc builds seems to override the results of the other? In that case that's a bug in those jobs. Probably related to force-pushing, but I can't see your code.

**Reply by cadop** [reply]
Thanks so much for taking the time to explain this to me. 

I think you have pointed out well my lack of skill on this.  What you describe makes sense, I have two jobs each should build some docs, then some other thing takes those docs and deploys... but I wasn't able to figure out how to do that, so I went with just trying to match the same commands/process I use in a terminal, just in a `.yml` file.  

Perhaps it is the last point you made that this is happening because of force-pushing.  I don't think it was overriding the code, but it seems like the commit never happens. For example this is the commit history https://github.com/cadop/dhart/commits/docs  The first one is a change to python docs, but thats because I manually went to the action that was cancelled and re-ran it.  The commits before it should have changed both python and C# docs, but the commits are just C#.  (c# is changing and not c++ just because c# has timestamps, but each time I was trying to fix the python docs and wondering why they weren't getting updated). 

Its also likely I am not able to properly understand what is going on so the thing I am focused on is not what is causing the problem.  For reference, this is the yml file https://github.com/cadop/dhart/blob/main/.github/workflows/documentation.yml

**Reply by airtower-luna** [reply]
Yep, that workflow confirms my suspicion about force pushing (`git push -f`). So what happens there is:

1. The two jobs start at about the same time, and fetch the (then) current state of your repository. (By the way you have two `actions/checkout` invocations in each, you only need one each, the second deletes what the first one fetched.)
2. Both do their building.
3. One of the jobs finishes building first and commits its results to the `docs` branch. The commit gets pushed.
4. The second job finishes building and commits its results to the `docs` branch. But: It commits on top of the state of `docs` from step 1! The resulting `docs` branch is force-pushed, which means the commit from step 3 is gone.

That's a classic parallelization problem, all kinds of data structures can break if parallel reads and writes mix in the wrong way.

You'll need to synchronize the committing in some way. What I'd suggest is *not* to commit in those jobs. [Create artifacts of the documentation instead and gather them up in a third job](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts#passing-data-between-jobs-in-a-workflow). Then you can either commit everything at once (without needing the `fetch-depth: 0` or `git worktree` stuff!), or use the custom Pages workflow to upload directly, without needing the `docs` branch at all. :slightly_smiling_face:

**Reply by cadop** [reply]
Thank you so much for sticking with me to explain it.  I was pretty lost, thinking the max-parallel feature being at 0 should have stopped the concurrent runs, and because I could switch the platform from windows and linux independently between the jobs, I was thinking of them like totally separate things. 

That link is really helpful, I will try to make an artifact (its always confused me :) )
